amoro-ams/amoro-ams-server/src/test/java/org/apache/amoro/server/manager/TestKubernetesOptimizerContainer.java

```diff
@@ -0,0 +1,171 @@
+        ObjectMapper mapper = new ObjectMapper();
+        containerProperties.putAll(mapper.convertValue(container.get("properties"), Map.class));
+      }
+    }
+  }
+
+  private static String checkAndGetProperty(Map<String, String> properties, String key) {
+    Preconditions.checkState(
+        properties != null && properties.containsKey(key), "Cannot find %s in properties", key);
+    return properties.get(key);
+  }
+
+  @Test
+  public void testBuildPodTemplateConfig() {
+    Map<String, String> groupProperties = Maps.newHashMap();
+    groupProperties.putAll(this.containerProperties);
+
+    // before parameter merging
+    PodTemplate podTemplate =
+        kubernetesOptimizerContainer.initPodTemplateFromLocal(groupProperties);
+
+    // after parameter merging
+
+    // generating frontend
+    ResourceType resourceType = ResourceType.OPTIMIZER;
+    Map<String, String> properties = Maps.newHashMap();
+    properties.put("memory", "1024");
+    Resource resource =
+        new Resource.Builder("KubernetesContainer", "k8s", resourceType)
+            .setMemoryMb(0)
+            .setThreadCount(1)
+            .setProperties(properties)
+            .build();
+    groupProperties.putAll(resource.getProperties());
+
+    // generate pod start args
+    long memoryPerThread = Long.parseLong(checkAndGetProperty(groupProperties, MEMORY_PROPERTY));
+    long memory = memoryPerThread * resource.getThreadCount();
+    String startUpArgs =
+        String.format(
+            "/entrypoint.sh optimizer %s %s",
+            memory, kubernetesOptimizerContainer.buildOptimizerStartupArgsString(resource));
+    String image = checkAndGetProperty(groupProperties, IMAGE);
+    String pullPolicy = checkAndGetProperty(groupProperties, PULL_POLICY);
+    String pullSecrets = groupProperties.getOrDefault(PULL_SECRETS, "");
+    String cpuLimitFactorString = groupProperties.getOrDefault(CPU_FACTOR_PROPERTY, "1.0");
+    double cpuLimitFactor = Double.parseDouble(cpuLimitFactorString);
+    int cpuLimit = (int) (Math.ceil(cpuLimitFactor * resource.getThreadCount()));
+
+    List<LocalObjectReference> imagePullSecretsList =
+        Arrays.stream(pullSecrets.split(";"))
+            .map(secret -> new LocalObjectReferenceBuilder().withName(secret).build())
+            .collect(Collectors.toList());
+
+    String resourceId = resource.getResourceId();
+    String groupName = resource.getGroupName();
+
+    Deployment deployment =
+        kubernetesOptimizerContainer.initPodTemplateFromFrontEnd(
+            podTemplate,
+            image,
+            pullPolicy,
+            cpuLimit,
+            groupName,
+            resourceId,
+            startUpArgs,
+            memory,
+            imagePullSecretsList);
+
+    Assert.assertEquals("amoro-optimizer-" + resourceId, deployment.getMetadata().getName());
+    Assert.assertEquals(
+        "k8s",
+        deployment.getSpec().getTemplate().getMetadata().getLabels().get("AmoroOptimizerGroup"));
```

Comment:
Here they need to reflect their mutual priorities and need to be split into different methods for independent testing to cover each other's priorities.